File: get_code.py
# https://github.com/rdassignies/pylegifrance
import os 
from dotenv import load_dotenv

load_dotenv()
# from pylegifrance import LegiHandler, recherche_CODE
# client = LegiHandler()
# client.set_api_keys(
#         legifrance_api_key=os.getenv("LEGIFRANCE_API_KEY"),
#         legifrance_api_secret=os.getenv("LEGIFRANCE_API_SECRET")
#         )

# art7 = recherche_CODE(code_name="Code civil", search="7")
# print(art7)

# https://dev.to/h4c5/scraping-de-legifrance-considerez-plutot-lapi-5afi

import requests
from requests_oauthlib import OAuth2Session




client_id = os.getenv("LEGIFRANCE_CLIENT_ID")
client_secret = os.getenv("LEGIFRANCE_CLIENT_SECRET")
# url_piste = "https://oauth.piste.gouv.fr/api/oauth/token"
url_piste = "https://sandbox-oauth.piste.gouv.fr/api/oauth/token"
# ok sur sandbox avec les client_id et client_secret de la premiere appli

# print(client_id)
# print(client_secret)

res = requests.post(
  url_piste,
  data={
    "grant_type": "client_credentials",
    "client_id": client_id,
    "client_secret": client_secret,
    "scope": "openid"
  }
)

token = res.json()
print(token)

# client = OAuth2Session(client_id, token=token)

# art = client.get("https://api.piste.gouv.fr/dila/legifrance/lf-engine-app/consult/getArticle", params={"id": "LEGIARTI000006307920"})
# print(art)
# import requests
# auth = "Bearer " + token["access_token"]

# response = requests.post(
#     "https://sandbox-api.piste.gouv.fr/dila/legifrance/lf-engine-app/consult/code",
#     json={"id": "LEGITEXT000006069577"},
#     headers={
#         "Authorization": auth,
#         "Content-Type": "application/json",
#     },
# )

# code =response.json()
# print(code)

# auth_token='kbkcmbkcmbkcbc9ic9vixc9vixc9v'
headers = {'Authorization': f'Bearer {token["access_token"]}', 'Content-Type': 'application/json', 'accept': 'application/json'}

url_ping = 'https://sandbox-api.piste.gouv.fr/dila/legifrance/lf-engine-app/list/ping'
ping = requests.get(url_ping, headers=headers)
print('PING',ping)

# data = {'app' : 'aaaaa'}
data ={
  "abrogated": True,
  "textId": "LEGITEXT000006075116",
  "searchedString": "constitution 1958",
  "date": "2021-04-15",
  "fromSuggest": True,
  "sctCid": "LEGISCTA000006112861"
}

# url = 'https://api.xy.com'
root_url = 'https://sandbox-api.piste.gouv.fr'
base_url='/dila/legifrance/lf-engine-app'
url = 'https://sandbox-api.piste.gouv.fr/dila/legifrance/lf-engine-app/consult/code'
response = requests.post(url, json=data, headers=headers)
print(response)
print(response.json())

File: get_code.js
require('dotenv').config(); 
// https://auth0.com/docs/get-started/authentication-and-authorization-flow/client-credentials-flow/call-your-api-using-the-client-credentials-flow


// const request = require('request');

// body={
//     "grant_type": "client_credentials",
//     "client_id": process.env.LEGIFRANCE_CLIENT_ID,
//     "client_secret": process.env.LEGIFRANCE_CLIENT_SECRET,
//     "scope": "openid"
//   }

//   console.log(body);

// request.post({
//     // headers: {'content-type' : 'application/x-www-form-urlencoded'},
//     url:     "https://sandbox-oauth.piste.gouv.fr/api/oauth/token",
//     body:    JSON.stringify(body), //body
//   }, function(error, response, body){
//     console.log(body);
//   });



var axios = require("axios").default;

var options = {
  method: 'POST',
//   url: 'https://{yourDomain}/oauth/token',
url : "https://sandbox-oauth.piste.gouv.fr/api/oauth/token",
//   headers: {'content-type': 'application/x-www-form-urlencoded'},
  data: new URLSearchParams({
    grant_type: 'client_credentials',
    client_id: process.env.LEGIFRANCE_CLIENT_ID,
    client_secret: process.env.LEGIFRANCE_CLIENT_SECRET,
    // audience: 'YOUR_API_IDENTIFIER'
"scope": "openid"
})
};

axios.request(options).then(function (response) {
  console.log(response.data);

//   let body = {
//     "recherche": {
//         "champs": [
//             {
//                 "typeChamp": "NUM_ARTICLE",
//                 "criteres": [
//                     {
//                         "typeRecherche": "EXACTE",
//                         "valeur": "L36-11",
//                         "operateur": "ET"
//                     }
//                 ],
//                 "operateur": "ET"
//             }
//         ],
//         "filtres": [
//             {
//                 "facette": "NOM_CODE",
//                 "valeurs": [
//                     "Code des postes et des communications électroniques"
//                 ]
//             },
//             {
//                 "facette": "DATE_VERSION",
//                 "singleDate": 1514802418000
//             }
//         ],
//         "pageNumber": 1,
//         "pageSize": 10,
//         "operateur": "ET",
//     "sort": "PERTINENCE",
//         "typePagination": "ARTICLE"
//     },
//    "fond": "CODE_DATE"
// }

// let body = {"id": "LEGITEXT000006069577"}


// BODYok
// let body = {
//     "abrogated": true,
//     "textId": "LEGITEXT000006075116",
//     "searchedString": "constitution 1958",
//     "date": "2021-04-15",
//     "fromSuggest": true,
//     "sctCid": "LEGISCTA000006112861"
//   }
//   let body = {
//     "abrogated": true,
//     "textId": "LEGITEXT000006075116",
//     // "searchedString": "constitution 1958",
//     "date": "2021-04-15",
//     // "fromSuggest": true,
//     "sctCid": "LEGISCTA000006112861"
//   }

  let body = {
    "abrogated": false,
    "textId": "LEGITEXT000006069577",
    "date": "2025-04-15",
    "sctCid": "LEGISCTA000006084232"
  }

const token = response.data.access_token;
  var options = {
    method: 'POST',
    // url: 'https://sandbox-api.piste.gouv.fr/dila/legifrance/lf-engine-app/list/ping',
    // url: 'https://sandbox-api.piste.gouv.fr/dila/legifrance/lf-engine-app/search',

url: "https://sandbox-api.piste.gouv.fr/dila/legifrance/lf-engine-app/consult/code",
    headers: {'content-type': 'application/json', authorization: 'Bearer '+token, 'accept': 'application/json'},
    data: body
  };
  
  axios.request(options).then(function (response) {
    console.log(response.data);
    let sections = response.data.sections;
    for (let i = 0; i < sections.length; i++) {
      console.log("\n########\n",sections[i].id, sections[i].title);
      if(sections[i].sections.length>0){
        console.log("______section_0_____",sections[i].sections[0]);
      }
      if(sections[i].articles.length>0){
        console.log("______article_0_____",sections[i].articles[0]);
      }
      
    }

  }).catch(function (error) {
    console.error(error);
  });
}).catch(function (error) {
  console.error(error);
});




// const options = {
// 	url: 'https://sandbox-api.piste.gouv.fr/dila/legifrance/lf-engine-app/list/ping',
// 	method: 'GET',
// 	headers: {
// 		'accept': 'text/plain'
// 	},
// 	gzip: true
// };
// request(options, (err, response, body) => {
//     console.log(response.statusCode, response);
// 	console.log('success:', body);
// });